language-server/src/main/kotlin/tools/samt/ls/SamtWorkspace.kt

```diff
@@ -1,40 +1,93 @@
 package tools.samt.ls
 
-import tools.samt.common.DiagnosticController
+import org.eclipse.lsp4j.PublishDiagnosticsParams
+import org.eclipse.lsp4j.services.LanguageClient
 import tools.samt.common.DiagnosticMessage
 import tools.samt.semantic.Package
-import tools.samt.semantic.SemanticModelBuilder
 import java.net.URI
 
-class SamtWorkspace(private val parserController: DiagnosticController) : Iterable<FileInfo> {
-    private val files = mutableMapOf<URI, FileInfo>()
-    var samtPackage: Package? = null
-        private set
-    private var semanticController: DiagnosticController =
-        DiagnosticController(parserController.workingDirectory)
+class SamtWorkspace {
+    private val folders = mutableMapOf<URI, SamtFolder>()
+    private val changedFolders = mutableSetOf<SamtFolder>()
+    private val removedFiles = mutableSetOf<URI>()
 
-    fun add(fileInfo: FileInfo) {
-        files[fileInfo.sourceFile.path] = fileInfo
+    fun getFolderSnapshot(path: URI): FolderSnapshot? = getFolder(path)?.let { FolderSnapshot(it.path, it.toList(), it.globalPackage) }
+
+    fun addFolder(folder: SamtFolder) {
+        val newPath = folder.path
+        // folder is contained in other folder, ignore
+        if (folders.keys.any { newPath.startsWith(it) }) {
+            return
+        }
+        // remove folders contained in new folder
+        folders.keys.removeIf { it.startsWith(newPath) }
+        folders[folder.path] = folder
+        changedFolders.add(folder)
+    }
+
+    fun removeFolder(path: URI) {
+        folders.remove(path)?.let { folder ->
+            removedFiles.addAll(folder.map { it.path })
+        }
     }
 
-    operator fun get(path: URI): FileInfo? = files[path]
+    fun getFile(path: URI): FileInfo? = getFolder(path)?.get(path)
 
-    override fun iterator(): Iterator<FileInfo> = files.values.iterator()
+    fun setFile(file: FileInfo) {
+        val folder = getFolder(file.path) ?: return
+        folder.set(file)
```

Comment:
We could compare the content of the file to make sure it actually changed to skip future calculations. This would make sense if we run through this code multiple times due to the different watchers we now have, otherwise it's probably a premature optimization